### PR TITLE
PHP 7 on Travis-CI (re-redux)

### PR DESCRIPTION
**Pull request by weierophinney**
This is a version of #7270 (which was in turn a rebased version of #7263), rebased against master, to support PHP 7.

I have tested both against a nightly version from 2 weeks ago, as well as the Coercive STH patch. The latterwhich provided me with a number of issues in the ZF2 codebase with regard to passing invalid arguments to internal functions, and the bulk of the commits in this branch address those, keeping compatibility with PHP 5 versions as well as PHP 7 nightly.

There is one issue I've found in `Zend\Stdlib\ArrayUtils::filter()`: apparently, PHP 7 nightly has a regression in its `array_filter` implementation; the third argument, which indicates whether to pass values only, keys only, or both to the composed callback, is ignored, causing one test case to fail.

Otherwise, this version should be ready to merge anytime.

### CODE REVIEW

**Comment by weierophinney** [comment]
Only failure on PHP 7 at this point is with `Zend\Stdlib\ArrayUtils::filter()`, which is due to a regression in PHP 7's `array_filter()` implementation, which I have reported here: https://bugs.php.net/bug.php?id=69299

**Comment by weierophinney** [comment]
This change will require some documentation:

- `Zend\Filter\Int` is deprecated (`int` is a reserved keyword in PHP 7); `Zend\Filter\ToInt` is preferred.
- `Zend\I18n\Validator\Int` is deprecated; `Zend\I18n\Validator\IsInt` is preferred.
- `Zend\I18n\Validator\Float` is deprecated (`float` is a reserved keyword in PHP 7); `Zend\I18n\Valdidator\IsFloat` is preferred.
- `Zend\Db\Sql\Ddl\Column\Float` is deprecated; `Zend\Db\Sql\Ddl\Column\Floating` is preferred.
- `Zend\XmlRpc\Value\String` has been renamed to `Zend\XmlRpc\Value\Text` (`string` is a reserved keyword in PHP 7). This should only affect developers who are directly instantiating XML-RPC value objects, which is so uncommon as to not warrant a BC shim.

For filters and validators, if you are pulling them from the plugin managers, you will continue to be able to do so under the old names, and the new class will be resolved; if you are instantiating them directly, you will receive a deprecation notice indicating the new class to use.

File: library/Zend/Cache/Storage/Adapter/Dba.php [hunk 1]
```diff
@@ -377,8 +378,10 @@ protected function internalSetItem(& $normalizedKey, & $value)
         $prefix      = ($namespace === '') ? '' : $namespace . $options->getNamespaceSeparator();
         $internalKey = $prefix . $normalizedKey;
 
+        $cacheableValue = (string) $value; // dba_replace requires a string
+
         $this->_open();
-        if (!dba_replace($internalKey, $value, $this->handle)) {
+        if (!dba_replace($internalKey, $cacheableValue, $this->handle)) {
```
Comment (DASPRiD) [inline library/Zend/Cache/Storage/Adapter/Dba.php]: This variable is only used once, why not cast it in-place?

Comment (weierophinney) [inline library/Zend/Cache/Storage/Adapter/Dba.php]: So I could provide a comment as to why the casting was happening.

**Comment by weierophinney** [comment]
all is well! As @DASPRiD has reviewed, I'll merge, and push an RC3 immediately.

**Comment by weierophinney** [comment]
So, @DASPRiD pointed out two more PHP RFCs that would affect this:

- https://wiki.php.net/rfc/reserve_more_types_in_php_7
- https://wiki.php.net/rfc/reserve_even_more_types_in_php_7

I found 5 more classes that will need shims for forward and backwards compat:

- `Zend\Form\Annotation\Object`
- `Zend\Filter\Null`
- `Zend\Paginator\Adapter\Null`
- `Zend\Mail\Transport\Null`

I'll work on these now. Until those are ready, I'm marking this as WIP.

**Comment by gianarb** [comment]
:+1: Thanks!!

**Pull request merged**